### ISSUE TITLE
Is there a way to remove Faraday's User-Agent header as a default option for all connections?

### ISSUE THREAD

**Issue by dmytro-savochkin** [lostisland/faraday#1248]
## Basic Info

* Faraday Version: 1.3.0
* Ruby Version: 2.7.2

## Issue description
Is there a way to prevent Faraday from sending its User-Agent header (`Faraday v1.3.0`)? I need it because one of the external servers we are dependent upon has banned it (https://github.com/lostisland/faraday/issues/1247#issuecomment-799345338).
Well, I know how to rewrite this header with something else per connection. Also I found how to rewrite it with something else as a default but only when there are no headers added when creating a connection (the new headers overwrite the default ones):

```
Faraday.default_connection_options = { headers: { user_agent: 'Something 1.2.3' } }
headers = { 'CustomHeader' => 'CustomValue' }
connection = Faraday.new('https://www.google.com', headers: headers) do |f|
  f.response :logger, nil, { headers: true }
end
connection.get

I, [2021-03-15T13:41:49.910701 #7925]  INFO -- request: GET https://www.google.com/
I, [2021-03-15T13:41:49.910993 #7925]  INFO -- request: CustomHeader: "CustomValue" User-Agent: "Faraday v1.3.0"
```

But I cannot find a way as how to remove that header completely (except for monkey-patching of course). Is there one? 

Thank you!

**Comment by iMacTia** [comment]
Hi @dmytro-savochkin, you're right this is not currently possible because passing the `headers` option to the connection will override any `headers` options set in `Faraday.default_connection_options`.
So as you said, a temporary solution would be to override the header on each connection, but I appreciate this is not ideal when you have many of them (although it's rarely a good idea to have many connections, to be honest).

In order to solve this on a global level, I can see 2 possible solutions:

1. We change the way we merge connection options with default_connection_options by performing a `deep_merge` (there's a helper in `Faraday::Utils` so we don't need to pull `ActiveSupport` in). This will solve the problem as `headers` won't be replaced, but merged gracefully. I'd very much prefer this solution, but I'm unsure on the backwards-compatibility impact
2. We introduce a new `Faraday.default_user_agent` setting that will default to the current value, but can also be overridden and, if necessary, even set to `nil` to remove the header completely.

@olleolleolle your input on the above?

**Comment by dmytro-savochkin** [comment]
Just FYI, I ended monkey-patching Faraday for now to stop it from setting a User-Agent value.
However, after doing it I found that Net::HTTP is doing absolutely the same! So I actually switched to excon adapter, which allows its users to either set a User-Agent or remove it completely with something like `Excon.defaults[:headers].delete('User-Agent')`.

**Comment by iMacTia** [comment]
Thanks for getting back to us @dmytro-savochkin and sorry such. simple thing ended up being so complicated!
I'm glad you solved your issue but I'll leave this one open in case anyone wants to submit a PR to add this feature 👍

**Comment by iMacTia** [comment]
This was fixed in #1343, thank you @xkwd 🙏!
